### PR TITLE
Adding rounded borders awareness for bevy_ui

### PR DESCRIPTION
**Pull request by rlofc**
Needed to implement this for an experiment using bevy_ui. Might be useful.

### CODE REVIEW

**Comment by atbentley** [comment]
Thanks for the PR! I like that you've used signed distances here, I wanted to take things in that direction, so this is a nice start. I'll take a proper look at this early next week and try to get it merged in.

File: src/bevy_ui.rs [hunk 1]
```diff
@@ -32,13 +33,22 @@ pub fn compute_blur_regions<const N: usize>(
...
                 node.size() * window.scale_factor(),
             );
-            blur_regions.blur(region);
+            let resolved = [
+                border_radius.top_left,
+                border_radius.top_right,
+                border_radius.bottom_right,
+                border_radius.bottom_left,
+            ]
+            .map(|v| v.resolve(node.size().y, viewport_size).unwrap_or(0.0));
```
Comment (atbentley) [inline src/bevy_ui.rs]:
```suggestion
            .map(|v| v.resolve(node.size().y, viewport_size).unwrap_or(0.0) * window.scale_factor());
```

Resolving the Val returns logical pixels, but the shader expects physical pixels.

Without converting to physical the blur comes outside of the radius, check out the bottom right corner in this screenshot:
Screenshot 2024-09-17 at 20 51 09

**Review by atbentley — COMMENTED** [review]
I'm on board with this PR, I would be happy to merge it in once the issues are fixed. Are you happy to look at it further or do you want me to pick it up from here?

Apart from the physical/logical issue, it looks like when there are two blur regions, only the first region is actually blurred. You can view this in the egui example, or modify the bevy_ui example:

Screenshot 2024-09-17 at 21 25 43

I tried to see what was causing it but I'm out of time tonight.

**Comment by rlofc** [comment]
Thanks for pointing out the logical/physical issue - learned a new thing. The multi-region support was a silly bug in the shader. Both should be fixed in the added commit.

File: src/bevy_ui.rs [hunk 2]
```diff
@@ -32,13 +33,22 @@ pub fn compute_blur_regions<const N: usize>(
...
                 node.size() * window.scale_factor(),
             );
-            blur_regions.blur(region);
+            let resolved = [
+                border_radius.top_left,
+                border_radius.top_right,
+                border_radius.bottom_right,
+                border_radius.bottom_left,
+            ]
+            .map(|v| v.resolve(node.size().y, viewport_size).unwrap_or(0.0) * window.scale_factor());
```
Comment (atbentley) [inline src/bevy_ui.rs:50]:
```suggestion
            .map(|v| v.resolve(node.size().y, viewport_size).unwrap_or(0.0) * SQRT_2 * window.scale_factor());
```

**Comment by atbentley** [comment]
I was looking at the bevy source for the ui material, and the logical to physical conversion isn't being done there. I'm not sure if sampling the 3D camera's output forced me down this path, or if I'm doing something that is unnecessary.

**Review by atbentley — COMMENTED** [review]
I was looking at this some more. I set the shader to output solid red and it seems that this shader still isn't matching the bevy shader perfectly.

Screenshot 2024-09-18 at 20 32 53

I traced this shader's radius and the bevy ui shader's radius in affinitiy and it was off by a factor of about 1.41 which is close to sqrt(2). Adding that extra term in makes everything line up perfectly. I have no idea why a factor of sqrt(2) is needed here.

I also quickly added some egui support for testing there and it also seems to work well. I did notice however that if the border radius exceeds half the smallest dimension our behaviour deviates from bevy ui / egui. But I'm happy to leave it out of this PR. Bevy UI is the top left in this recording:

https://github.com/user-attachments/assets/7775403e-2e78-470b-9bf3-7a10c6c909cb

**Comment by atbentley** [comment]
Turns out the odd reason why the sqrt(2) was needed is because I messed things up while trying to add the egui support 😅.

So turns out this works fine!

**Pull request merged**